tests/test_filtering.py

```diff
@@ -0,0 +1,251 @@
+        "err2l": lambda x: np.maximum(np.mean(x) - np.percentile(x, 2.5), 0),
+    }
+
+    for stat in stats:
+        spstatkey = statistic_dict[stat] if stat in statistic_dict.keys() else stat
+        result[stat], _, _ = binstat(x_in, y_in, bins=bins, statistic=spstatkey)
+
+    return result
+
+
+@pytest.mark.parametrize("R", R_PARAM_LIST)
+@pytest.mark.parametrize("filter_flag", options_filter)
+def test_filters(filter_flag, R, plt):
+    opts = prd.get_all_options(redshift=10.0)
+
+    up = UserParams(opts["user_params"])
+    cp = CosmoParams(opts["cosmo_params"])
+    ap = AstroParams(opts["astro_params"])
+    fo = FlagOptions(opts["flag_options"])
+
+    # testing a single pixel source
+    input_box_centre = np.zeros((up.HII_DIM,) * 3, dtype="f4")
+    input_box_centre[up.HII_DIM // 2, up.HII_DIM // 2, up.HII_DIM // 2] = 1.0
+    output_box_centre = np.zeros((up.HII_DIM,) * 3, dtype="f8")
+    # use MFP=20 for the exp filter, use a 3 cell shell for the annular filter
+    if filter_flag == 3:
+        R_param = 20
+    elif filter_flag == 4:
+        R_param = max(R - 3 * (up.BOX_LEN / up.HII_DIM), 0)
+    else:
+        R_param = 0
+
+    lib.test_filter(
+        up(),
+        cp(),
+        ap(),
+        fo(),
+        ffi.cast("float *", input_box_centre.ctypes.data),
+        R,
+        R_param,
+        filter_flag,
+        ffi.cast("double *", output_box_centre.ctypes.data),
+    )
+
+    # expected outputs given in cell units
+    R_cells = R / up.BOX_LEN * up.HII_DIM
+    Rp_cells = R_param / up.BOX_LEN * up.HII_DIM
+    r_from_centre = np.linalg.norm(
+        np.mgrid[0 : up.HII_DIM, 0 : up.HII_DIM, 0 : up.HII_DIM]
+        - np.array(
+            [
+                up.HII_DIM // 2,
+                up.HII_DIM // 2,
+                up.HII_DIM // 2,
+            ]
+        )[:, None, None, None],
+        axis=0,
+    )
+    # average value of R in a sphere with same volume of the cell
+    r_from_centre[up.HII_DIM // 2, up.HII_DIM // 2, up.HII_DIM // 2] = (
+        np.pi * (0.62035) ** 4
+    )
+    exp_output_centre = get_expected_output_centre(
+        r_from_centre, R_cells, Rp_cells, filter_flag
+    )
+
+    if plt == mpl.pyplot:
+        r_bins = np.linspace(0, (up.HII_DIM / 2 * np.sqrt(3)), num=32)
+        r_cen = (r_bins[1:] + r_bins[:-1]) / 2
+        binned_truth_centre = get_expected_output_centre(
+            r_cen, R_cells, Rp_cells, filter_flag
+        )
+        filter_plot(
+            inputs=[
+                input_box_centre,
+            ],
+            outputs=[
+                output_box_centre,
+            ],
+            binned_truths=[
+                binned_truth_centre,
+            ],
+            truths=[
+                exp_output_centre,
+            ],
+            r_bins=r_bins,
+            r_grid=r_from_centre,
+            slice_index=up.HII_DIM // 2,
+            slice_axis=0,
+            plt=plt,
```

Comment:
Something weird happened with the formatting here -- leave off the trailing commas within the lists so that they stay on one line.